File: solver/src/test/java/will/sudoku/solver/PuzzleGeneratorTest.kt
package will.sudoku.solver

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

/**
 * Test for PuzzleGenerator.
 *
 * Tests puzzle generation and validity.
 */
class PuzzleGeneratorTest {

    /** Solver for verifying generated puzzles — uses basic eliminators
     *  to avoid OOM on resource-constrained CI. A puzzle solvable by
     *  basic techniques is still a valid puzzle. */
    private val verifier = Solver(SolverConfig.basic())

    @Test
    fun `generates solvable puzzle`() {
        val puzzle = PuzzleGenerator.generate()

        // Puzzle should be solvable
        val solution = verifier.solve(puzzle)

        assertThat(solution).isNotNull
        assertThat(solution!!.isSolved()).isTrue()
    }

    @Test
    fun `generates puzzle with reproducible seed`() {
        val puzzle1 = PuzzleGenerator.generate(seed = 12345)
        val puzzle2 = PuzzleGenerator.generate(seed = 12345)

        // Same seed should produce same puzzle
        assertThat(puzzle1).isEqualTo(puzzle2)
    }

    @Test
    fun `generates different puzzles with different seeds`() {
        val puzzle1 = PuzzleGenerator.generate(seed = 11111)
        val puzzle2 = PuzzleGenerator.generate(seed = 22222)

        // Different seeds should produce different puzzles (usually)
        assertThat(puzzle1).isNotEqualTo(puzzle2)
    }

    @Test
    fun `generates easy puzzle with more clues`() {
        val easyPuzzle = PuzzleGenerator.generate(DifficultyRater.Level.EASY)

        // Count filled cells
        val filledCount = Coord.all.count { easyPuzzle.isConfirmed(it) }

        // Easy puzzles should have more clues (around 46)
        assertThat(filledCount).isGreaterThanOrEqualTo(40)
    }

    @Test
    fun `generates hard puzzle with fewer clues`() {
        val hardPuzzle = PuzzleGenerator.generate(DifficultyRater.Level.EXPERT)

        // Count filled cells
        val filledCount = Coord.all.count { hardPuzzle.isConfirmed(it) }

        // Expert puzzles should have fewer clues (around 25)
        assertThat(filledCount).isGreaterThanOrEqualTo(20)
    }

    @Test
    fun `generated puzzle is valid`() {
        val puzzle = PuzzleGenerator.generate()

        // Puzzle should be in a valid state
        assertThat(puzzle.isValid()).isTrue()
    }

    @Test
    fun `generated puzzle is not solved`() {
        val puzzle = PuzzleGenerator.generate()

        // Generated puzzle should not be already solved
        assertThat(puzzle.isSolved()).isFalse()
    }

    @Test
    fun `generates minimal puzzle`() {
        // Use fixed seed for deterministic test
        val puzzle = PuzzleGenerator.generateMinimal(seed = 12345)

        // Minimal puzzle should be solvable
        val solution = verifier.solve(puzzle)

        assertThat(solution).isNotNull
        assertThat(solution!!.isSolved()).isTrue()
    }
}

File: packages/solver/tests/PuzzleGenerator.test.ts
import { describe, it, expect } from 'vitest'
import { Board } from '../src/Board'
import { Coord } from '../src/Coord'
import { Solver } from '../src/Solver'
import { Level } from '../src/DifficultyRater'
import { generate } from '../src/PuzzleGenerator'

describe('PuzzleGenerator', () => {
  it('generates a valid easy puzzle', () => {
    const puzzle = generate(Level.EASY, 42)
    const solver = new Solver()
    const solution = solver.solve(puzzle.copy())
    expect(solution).not.toBeNull()
  }, 15000)

  it('generates a valid medium puzzle', () => {
    const puzzle = generate(Level.MEDIUM, 42)
    const solver = new Solver()
    const solution = solver.solve(puzzle.copy())
    expect(solution).not.toBeNull()
  }, 15000)

  it('seed produces reproducible puzzles', () => {
    const p1 = generate(Level.MEDIUM, 123)
    const p2 = generate(Level.MEDIUM, 123)
    for (let i = 0; i < 81; i++) {
      expect(p1.candidatePatterns[i]).toBe(p2.candidatePatterns[i])
    }
  }, 15000)

  it('different seeds produce different puzzles', () => {
    const p1 = generate(Level.MEDIUM, 1)
    const p2 = generate(Level.MEDIUM, 2)
    const same = Array.from({ length: 81 }, (_, i) => p1.candidatePatterns[i] === p2.candidatePatterns[i])
      .every(Boolean)
    expect(same).toBe(false)
  }, 15000)

  it('generated puzzles have confirmed cells', () => {
    const puzzle = generate(Level.MEDIUM, 42)
    let confirmed = 0
    for (const c of Coord.all) {
      if (puzzle.isConfirmed(c)) confirmed++
    }
    // Medium should have ~36 cells (81 - 45)
    expect(confirmed).toBeGreaterThan(20)
    expect(confirmed).toBeLessThan(60)
  }, 15000)
})
